src/lib/asn1/asn1_oid.cpp

```diff
@@ -194,35 +210,73 @@ void OID::decode_from(BER_Decoder& decoder) {
       throw BER_Bad_Tag("Error decoding OID, unknown tag", obj.tagging());
    }
 
-   const size_t length = obj.length();
-   const uint8_t* bits = obj.bits();
-
-   if(length < 2 && !(length == 1 && bits[0] == 0)) {
+   if(obj.length() == 0) {
       throw BER_Decoding_Error("OID encoding is too short");
    }
 
-   m_id.clear();
-   m_id.push_back(bits[0] / 40);
-   m_id.push_back(bits[0] % 40);
+   auto consume = [](std::span<const uint8_t> data) -> std::pair<std::span<const uint8_t>, uint32_t> {
+      BOTAN_ASSERT_NOMSG(!data.empty());
 
-   size_t i = 0;
-   while(i != length - 1) {
-      uint32_t component = 0;
-      while(i != length - 1) {
-         ++i;
+      uint32_t b = data.front();
 
-         if(component >> (32 - 7)) {
-            throw Decoding_Error("OID component overflow");
+      if(b <= 0x7F) {
+         return std::make_pair(data.subspan(1), b);
+      } else {
+         b &= 0x7F;
+         // Even BER requires that the OID have minimal length, ie that
+         // the first byte of a multibyte encoding cannot be zero
+         // See X.690 section 8.19.2
+         if(b == 0) {
+            throw Decoding_Error("Leading zero byte in multibyte OID encoding");
          }
+         data = data.subspan(1);
+         while(!data.empty()) {
+            const auto next = data.front();
+            data = data.subspan(1);
+            const bool more = (next & 0x80);
+
+            if(b >> (32 - 7)) {
+               throw Decoding_Error("OID component overflow");
+            }
+            b <<= 7;
+            b |= (next & 0x7F);
 
-         component = (component << 7) + (bits[i] & 0x7F);
-
-         if(!(bits[i] & 0x80)) {
-            break;
+            if(!more) {
+               return std::make_pair(data, b);
+            }
          }
+         throw Decoding_Error("Truncated OID value");
+      }
+   };
+
+   std::span<const uint8_t> span(obj.bits(), obj.length());
+   std::vector<uint32_t> parts;
+   while(!span.empty()) {
+      auto [subspan, comp] = consume(span);
+      span = subspan;
```

Comment:
Here's the invocation of `consume` that uses `BufferSlicer&` (as suggested above):

```suggestion
   BufferSlicer data({obj.bits(), obj.length()});
   std::vector<uint32_t> parts;
   while(!data.empty()) {
      const auto comp = consume(data);
```

Reply:
:+1: